File: sdk/android/SignetSDK.kt
package ing.paperclip.signet

import android.graphics.Bitmap
import android.graphics.Color
import androidx.camera.core.ImageProxy
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * SignetSDK — drop this file into any Android camera app.
 *
 * Usage:
 *   1. Add libsignet.so to your jniLibs/ folder.
 *   2. Call SignetSDK.start() when your CameraX session starts.
 *      This begins prefetching the drand beacon every 25 seconds.
 *   3. In your ImageCapture.OnImageCapturedCallback or ImageAnalysis.Analyzer,
 *      call stamp(bitmap) on the raw Bitmap BEFORE encoding or saving.
 *   4. Save as PNG or JPEG quality >= 90.
 *
 * Verification:
 *   SignetSDK.verify(bitmap) returns a VerifyResult with the drand round
 *   and UTC timestamp, or null if no valid watermark is present.
 */
object SignetSDK {

    // Load the Signet native library
    init {
        System.loadLibrary("signet")
    }

    // ── Native declarations ───────────────────────────────────────────────────

    @JvmStatic private external fun prefetchRound(sigHexBuf: ByteArray): Long
    @JvmStatic private external fun stampPixels(
        pixelsRgb: ByteArray, width: Int, height: Int, sigHex: String
    ): Int
    @JvmStatic private external fun verifyPixels(
        pixelsRgb: ByteArray, width: Int, height: Int,
        outRound: LongArray, outUnixTime: LongArray
    ): Int

    // ── State ─────────────────────────────────────────────────────────────────

    @Volatile private var cachedSigHex: String? = null
    private val executor = Executors.newSingleThreadScheduledExecutor()
    private var prefetchJob: ScheduledFuture<*>? = null

    // ── Lifecycle ─────────────────────────────────────────────────────────────

    /** Start background beacon prefetch. Call when the camera session opens. */
    fun start() {
        prefetchJob = executor.scheduleAtFixedRate(::prefetchNow, 0, 25, TimeUnit.SECONDS)
    }

    /** Stop background prefetch. Call when the camera session closes. */
    fun stop() {
        prefetchJob?.cancel(false)
    }

    // ── Stamp ─────────────────────────────────────────────────────────────────

    /**
     * Stamp a Bitmap with the current drand watermark.
     *
     * Call this inside your photo capture pipeline, before encoding or saving.
     * The bitmap must be in ARGB_8888 format (Android default).
     *
     * Returns true if the stamp was applied. Returns false if no beacon is
     * cached yet (call start() before opening the camera).
     */
    fun stamp(bitmap: Bitmap): Boolean {
        val sig = cachedSigHex ?: return false
        val width = bitmap.width
        val height = bitmap.height
        val rgb = bitmapToRgb(bitmap)
        if (stampPixels(rgb, width, height, sig) != 0) return false
        rgbToBitmap(rgb, bitmap)
        return true
    }

    // ── Verify ────────────────────────────────────────────────────────────────

    data class VerifyResult(val round: Long, val unixTime: Long) {
        val isoTime: String get() {
            val date = java.util.Date(unixTime * 1000)
            return java.text.SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'",
                java.util.Locale.US).apply {
                timeZone = java.util.TimeZone.getTimeZone("UTC")
            }.format(date)
        }
    }

    /**
     * Verify the Signet watermark in a Bitmap.
     * Returns null if no valid watermark is found.
     */
    fun verify(bitmap: Bitmap): VerifyResult? {
        val rgb = bitmapToRgb(bitmap)
        val outRound = LongArray(1)
        val outTime = LongArray(1)
        val ok = verifyPixels(rgb, bitmap.width, bitmap.height, outRound, outTime)
        return if (ok == 1) VerifyResult(outRound[0], outTime[0]) else null
    }

    // ── Private ───────────────────────────────────────────────────────────────

    private fun prefetchNow() {
        val buf = ByteArray(512)
        val round = prefetchRound(buf)
        if (round > 0) {
            cachedSigHex = buf.takeWhile { it != 0.toByte() }
                .toByteArray().toString(Charsets.US_ASCII)
        }
    }

    private fun bitmapToRgb(bitmap: Bitmap): ByteArray {
        val w = bitmap.width
        val h = bitmap.height
        val buf = ByteArray(w * h * 3)
        val pixels = IntArray(w * h)
        bitmap.getPixels(pixels, 0, w, 0, 0, w, h)
        for (i in pixels.indices) {
            val p = pixels[i]
            buf[i * 3 + 0] = Color.red(p).toByte()
            buf[i * 3 + 1] = Color.green(p).toByte()
            buf[i * 3 + 2] = Color.blue(p).toByte()
        }
        return buf
    }

    private fun rgbToBitmap(rgb: ByteArray, bitmap: Bitmap) {
        val w = bitmap.width
        val h = bitmap.height
        val pixels = IntArray(w * h)
        bitmap.getPixels(pixels, 0, w, 0, 0, w, h)
        for (i in pixels.indices) {
            val r = rgb[i * 3 + 0].toInt() and 0xFF
            val g = rgb[i * 3 + 1].toInt() and 0xFF
            val b = rgb[i * 3 + 2].toInt() and 0xFF
            val a = Color.alpha(pixels[i])
            pixels[i] = Color.argb(a, r, g, b)
        }
        bitmap.setPixels(pixels, 0, w, 0, 0, w, h)
    }
}

File: sdk/ios/SignetSDK.swift
import Foundation
import CoreImage
import AVFoundation

// SignetSDK — drop this file into any iOS camera app.
//
// Usage:
//   1. Call SignetSDK.shared.start() when the camera session starts.
//      This begins prefetching the drand beacon every 25 seconds.
//   2. In your AVCapturePhotoCaptureDelegate or photo processing pipeline,
//      call stamp(pixelBuffer:) on the raw CVPixelBuffer BEFORE encoding.
//   3. Save the modified buffer as PNG (lossless) or JPEG at quality ≥ 0.9.
//
// Verification:
//   SignetSDK.verify(pixelBuffer:) returns a VerifyResult with the drand
//   round and UTC timestamp, or nil if the image has no valid watermark.

public final class SignetSDK {
    public static let shared = SignetSDK()

    private var cachedSigHex: String?
    private var cachedRound: UInt64 = 0
    private var prefetchTimer: Timer?
    private let sigBuf = UnsafeMutablePointer<CChar>.allocate(capacity: 512)

    private init() {}

    deinit {
        sigBuf.deallocate()
    }

    // MARK: - Lifecycle

    /// Start background beacon prefetch. Call when the camera session opens.
    public func start() {
        prefetchNow()
        prefetchTimer = Timer.scheduledTimer(withTimeInterval: 25, repeats: true) { [weak self] _ in
            self?.prefetchNow()
        }
    }

    /// Stop background prefetch. Call when the camera session closes.
    public func stop() {
        prefetchTimer?.invalidate()
        prefetchTimer = nil
    }

    // MARK: - Stamp

    /// Stamp a CVPixelBuffer with the current drand watermark.
    ///
    /// Call this inside your photo capture pipeline, on the raw pixels,
    /// before encoding or saving. The buffer must be kCVPixelFormatType_32BGRA
    /// or kCVPixelFormatType_24RGB.
    ///
    /// TODO: hardware attestation required for production. The device's public key
    /// certificate must be obtained from the Secure Enclave and passed to the C API
    /// so that verifiers can confirm the stamp came from a hardware-backed key.
    ///
    /// Returns true if the stamp was applied, false if no beacon is cached yet.
    @discardableResult
    public func stamp(pixelBuffer: CVPixelBuffer) -> Bool {
        guard let sig = cachedSigHex else { return false }
        CVPixelBufferLockBaseAddress(pixelBuffer, [])
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }

        guard let base = CVPixelBufferGetBaseAddress(pixelBuffer) else { return false }
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)

        // Convert BGRA → contiguous RGB scratch buffer, stamp, write back
        let npixels = width * height
        var rgb = [UInt8](repeating: 0, count: npixels * 3)
        let src = base.assumingMemoryBound(to: UInt8.self)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)

        for y in 0..<height {
            for x in 0..<width {
                let srcIdx = y * bytesPerRow + x * 4
                let dstIdx = (y * width + x) * 3
                rgb[dstIdx + 0] = src[srcIdx + 2] // R (from BGRA)
                rgb[dstIdx + 1] = src[srcIdx + 1] // G
                rgb[dstIdx + 2] = src[srcIdx + 0] // B
            }
        }

        let result = sig.withCString { sigPtr in
            signet_stamp_pixels(&rgb, Int32(width), Int32(height), sigPtr)
        }
        guard result == 0 else { return false }

        // Write stamped blue channel back into the pixel buffer
        for y in 0..<height {
            for x in 0..<width {
                let srcIdx = (y * width + x) * 3
                let dstIdx = y * bytesPerRow + x * 4
                src[dstIdx + 0] = rgb[srcIdx + 2] // B back to BGRA
            }
        }
        return true
    }

    // MARK: - Verify

    public struct VerifyResult {
        public let round: UInt64
        public let unixTime: UInt64
        /// Human-readable UTC timestamp
        public var dateString: String {
            let date = Date(timeIntervalSince1970: TimeInterval(unixTime))
            let fmt = ISO8601DateFormatter()
            return fmt.string(from: date)
        }
    }

    /// Verify the Signet watermark in a CVPixelBuffer.
    /// Returns nil if no valid watermark is found.
    public static func verify(pixelBuffer: CVPixelBuffer) -> VerifyResult? {
        CVPixelBufferLockBaseAddress(pixelBuffer, .readOnly)
        defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, .readOnly) }

        guard let base = CVPixelBufferGetBaseAddress(pixelBuffer) else { return nil }
        let width = CVPixelBufferGetWidth(pixelBuffer)
        let height = CVPixelBufferGetHeight(pixelBuffer)
        let bytesPerRow = CVPixelBufferGetBytesPerRow(pixelBuffer)
        let npixels = width * height
        var rgb = [UInt8](repeating: 0, count: npixels * 3)
        let src = base.assumingMemoryBound(to: UInt8.self)

        for y in 0..<height {
            for x in 0..<width {
                let srcIdx = y * bytesPerRow + x * 4
                let dstIdx = (y * width + x) * 3
                rgb[dstIdx + 0] = src[srcIdx + 2]
                rgb[dstIdx + 1] = src[srcIdx + 1]
                rgb[dstIdx + 2] = src[srcIdx + 0]
            }
        }

        var outRound: UInt64 = 0
        var outTime: UInt64 = 0
        let verified = signet_verify_pixels(&rgb, Int32(width), Int32(height), &outRound, &outTime)
        guard verified == 1 else { return nil }
        return VerifyResult(round: outRound, unixTime: outTime)
    }

    // MARK: - Private

    private func prefetchNow() {
        DispatchQueue.global(qos: .utility).async { [weak self] in
            guard let self else { return }
            var round: UInt64 = 0
            let ret = signet_prefetch_round(&round, self.sigBuf, 512)
            if ret == 0 {
                self.cachedRound = round
                self.cachedSigHex = String(cString: self.sigBuf)
            }
        }
    }
}
